README.md
![](https://img.shields.io/badge/version-v0.6.0-gold)  
![](https://img.shields.io/badge/python-v3.10.1-blue)
![](https://img.shields.io/badge/Flask-v2.1.3-pink)
![](https://img.shields.io/badge/sqlite3-v2.6.0-purple)
![](https://img.shields.io/badge/Docker-v20.10.17-orange)

![](https://img.shields.io/badge/pytest-v7.1.2-black)
![](https://img.shields.io/badge/passed_tests-12-brightgreen)
![](https://img.shields.io/badge/failed_tests-0-red)

![](https://img.shields.io/badge/coverage-100%25-brightgreen)

---

- [Solution Tree](#solution-tree)  
- [How To Run Locally](#how-to-run-locally)
- [How To Run With Docker](#how-to-run-with-docker)
- [Making Requests](#making-requests)
- [FYI](#fyi)
  - [Additional Information](#additional-information)
  - [What Did Go Right](#what-did-go-rigth)
  - [What Did Go Wrong](#what-did-go-wrong)
  - [Doubts](#doubts)

# Solution Tree

## The `app.py` module
The running process occurs in `app.py`. Module responsible for doing the entire Flask process.

As requested in the challenge, this API works with two methods. One `GET` and one `POST`. Both handled through the `/weather` route.
```
@app.route("/weather", methods=['GET', 'POST'])
```

## The `POST` method
Responsible for receiving a user defined id, and from that, making pre-defined requests to the Open Weather API.
Then store the response, so it can be consulted later.

After evaluating that the received request is a POST. The first thing the function does, is to get the `user_id`, from the body.
```
    if request.method == "POST":
        request_data = request.json

        try:
            user_id = request_data["user_id"]
        except KeyError:
            return {"message": "Missing param 'user_id'."}, 400
```
As we can see above, if the `user_id` param is not in the requested body, it will return an error.

After that, we call another function, responsible for validating the `user_id`. And only if the `user_id` is valid, we continue on the code.
```
if verify_user_id(user_id):
.
.
.
else:
    return {"message": "Invalid 'user_id'. It must contain only numbers and be unique."}, 400
```
>The `verify_user_id()` function checks every path possible for evaluating the `user_id`
> ```
> def verify_user_id(user_id):
>    try:
>        user_id = int(user_id)
>    except ValueError:
>        return False
>    except TypeError:
>        return False
>
>    sql = "SELECT user_id FROM weather"
>    db = db_connect()
>    cur = db.cursor()
>    cur.execute(sql)
>    rows = cur.fetchall()
>
>    for row in rows:
>        if user_id == row[0]:
>            return False
>
>    return True
> ```
> First, we try to convert it into an integer, because the `user_id` must contain only numbers, if it's not converted, we return `False`.
>
> Then, we query through de database, and check if there's already collected cities with this id (`user_id` is not unique, but it can only call API `POST` method once).

If everything is valid with the `user_id`, we can move on.

First, we get the request datetime, then generate a list with all the cities to collect. For doing so, we use another two functions `get_weather()` and `insert_weather()`
```
request_datetime = datetime.now()

cities = os.getenv("APPENDIX_A").split(',')
try:
    for city_id in cities:
        city_info = get_weather(city_id)
        insert_weather(user_id, request_datetime, str(city_info))
except Exception as e:
    return {"message": e}, 500
```
>The `get_weather()` function, calls the Open Weather API and returns the desired information.
> ```
> def get_weather(city_id):
>    weather_api = os.getenv('WEATHER_API')
>    api_key = os.getenv('API_KEY')
>
>    query = f"id={city_id}&appid={api_key}"
>
>    url = weather_api + query
>
>    data = requests.get(url).json()
>
>    return {
>        "city_id": data['id'],
>        "temp": data['main']['temp'],
>        "humidity": data['main']['humidity']
>    }
> ```
>
> The `insert_weather()` function picks all the information and insert them to the database
> ```
> def insert_weather(user_id, request_datetime, city_info):
>    sql = f'''
>    INSERT INTO weather(user_id, collected_at, city_info)
>    VALUES({user_id}, datetime("{request_datetime}"), ?)
>    '''
>
>    db = db_connect()
>    cur = db.cursor()
>    cur.execute(sql, (city_info,))
>    db.commit()
> ```

If any of this process go wrong, we return an error 500, because in this case, it must be a system error.

But if everything goes right we return a success message
```
return {"message": "All cities collected."}, 200
```

---

## The `GET` method
Responsible for receiving a `user_id`, and query through the database. Checking if the collecting method is completed. If not, we return the percentage of the process.

After evaluating that the received request is a `GET`. The first thing the function does, is to get the `user_id`, from the `url-query`.
```
if request.method == "GET":
    user_id = request.args.get("user-id")

    if not user_id:
        return {"message": "Missing query param 'user-id'."}, 400
```
The first thing that we check, is if the `user_id` param is present on the query url. If not, e return an error.

After that we check if the `user_id` is valid, by trying to convert it to an integer.
```
try:
    int(user_id)
except ValueError:
    return {"message": "Invalid 'user-id'. It must contain only numbers and be unique."}, 400
```

Then, we call `get_collected_cities()` function, which will try to get and return the cities caught by the POST request.
```
collected_cities = get_collected_cities(user_id)
```
> This method queries in database with the received `user_id`, and at first checks if it has rows. If not, we return a `False` statement.
> ```
> def get_collected_cities(user_id):
>    sql = f'''
>        SELECT collected_at, city_info FROM weather
>        WHERE user_id = {user_id}
>        '''
>
>    db = db_connect()
>    cur = db.cursor()
>    cur.execute(sql)
>    rows = cur.fetchall()
>
>    print(rows)
>    if len(rows) == 0:
>        return False
> ```
> If the query returns data, we compare the current number of collected cities, with the number of cities requested. And with this information, we can get the percentage.
> ```
>    collected_info = {}
>
>    cities_ids = os.getenv("APPENDIX_A").split(',')
>    progress = (len(rows) * 100) / len(cities_ids)
>
>    collected_info["user_id"] = user_id
>    collected_info["progress"] = f"{progress:.3g}%"
> ```
> Finishing, we get the information received from the database, and mount a `json object` for showing to the user.
> ```
>    cities = []
>    for row in rows:
>        city_info = row[1].replace("\'", "\"")
>        city_info = json.loads(city_info)
>        data_info = city_info
>        data_info["collected_at"] = row[0]
>
>        cities.append(data_info)
>
>    collected_info["cities"] = cities
>    return collected_info
> ```

If there were no data for the passed `user_id`, we return this error message, but if it has we return the object with the percentage and the cities collected.
```
if not collected_cities:
    return {"message": "User identifier not found."}, 400

return collected_cities, 200
```

# How To Run Locally
Requirements:
- [Git](https://git-scm.com/downloads)
- [Python3.10](https://www.python.org/downloads/)
- .env file containing the private access-token and the cities ids to Open Weather API _(you can ask this file to the project mantainer)_

1. Clone the repository  
`https://github.com/salatiel6/devgrid_challenge.git`


2. Open the challenge directory  
Widows/Linux:`cd devgrid_challenge`  
Mac: `open devgrig_challenge`


3. Create virtual environment (recommended)  
`python -m venv ./venv`


4. Activate virtual environment (recommended)  
Windows: `venv\Scripts\activate`  
Linux/Mac: `source venv/bin/activate`


5. Install every dependencies  
`pip install -r requirements.txt`


6. Open the tests' directory  
Windows/Linux: `cd tests`  
Mac: `open tests`


7. Run tests  
Without coverage: `pytest`  
With coverage: `pytest -vv --cov=. --cov-report=term-missing --cov-config=.coveragerc`


8. Get back at root directory
`cd ..`


9. Open the source directory  
Windows/Linux: `cd src`  
Mac: `open src`


10. Run the application  
`python app.py`

# How To Run With Docker
This application was developed in a Windows OS. So I'm not sure if Linux or Mac can run the container.

Requirements:
- [Git](https://git-scm.com/downloads)
- [Docker](https://www.docker.com/)
- .env file containing the private access-token and the cities ids to Open Weather API _(you can ask this file to the project mantainer)_

1. Clone the repository  
`https://github.com/salatiel6/devgrid_challenge.git`


2. Open the challenge directory  
Widows/Linux:`cd devgrid_challenge`  
Mac: `open devgrig_challenge`


3. Build docker image  
`docker-compose build`


4. Start docker container  
`docker-compose up -d`

# Making Requests
Recommended: 
- [Postman](https://www.postman.com/)  

After starting the application, with or without docker, it should be running at your `localhost`

## Endpoints

Search and insert weather:
> ![](https://img.shields.io/badge/method-POST-purple)  
> `url`: `http://localhost:5000/weather`  
>
> `body example`:
> ```
> {
>     "id": "123456"
> }
> ```

Get collected cities:
> ![](https://img.shields.io/badge/method-GET-pink)  
> 
> `query-url example`:  
> `url`: `http://localhost:5000/weather?id=123456`

**Note**: A valid `id` must contain only numbers and be unique when calling the POST method.

# FYI
## Additional Information
- `.env`: Configuration file that is not on the repository but is needed to run the application. Because it has the access token, and the cities ids to make request to Open Weather API. _(you can ask this file to the project mantainer)_  
- `pre-commit-config.yaml`: File which uses flake8 that is a python tool that glues together pycodestyle, pyflakes, mccabe, and third-party plugins to check the style and quality of some python code. It can be installed within git, so you can't commit anything out of the PEP 8 patterns.  
- `sqlite3`: Was chosen as the application database because of its simplicity and availability to run within Python and Docker.

## What Did Go Rigth
- Application runnig in local machine
- Application runnig in local docker image
- Database storing all the information requested
- Application and Database tests running with 100% pytest coverage

## What Did Go Wrong
- Couldn't split `app.py` in desired multiple modules(`run.py`, `views.py`, `features.py`). Because when running in Docker, it couldn't import these modules. It probably has something to do with PYTHONPATH, but I wasn't able to solve this problem.
- I tried to run pytest on Dockerfile, but got a little confused on how to manage the directories for manking it works.

## Doubts
- I don't know if I handled SQLite database files on propper way.
- The pytest coverage returns 100%. But I'm  not sure if everything is properly tested.
- Got some problems when trying to split app.py module, and I don't know what is the real problem.
- Not sure if it is a good pratice to run pytest on Dockerfile

ChangeLog.md
> ![](https://img.shields.io/badge/Jul_25,_2022-black) ![](https://img.shields.io/badge/v0.6.0-FEAT-blue)  
> `Documentation completed`

> ![](https://img.shields.io/badge/Jul_25,_2022-black) ![](https://img.shields.io/badge/v0.5.1-REFACTOR-red)  
> `Renamed all id reference to user_id`

> ![](https://img.shields.io/badge/Jul_25,_2022-black) ![](https://img.shields.io/badge/v0.5.0-FEAT-blue)  
> `Created method for checking if .env file exists`

> ![](https://img.shields.io/badge/Jul_25,_2022-black) ![](https://img.shields.io/badge/v0.4.2-REFACTOR-red)  
> `Removed unnecessary variables from .env. Renamed db_manager to test_db_manager`

> ![](https://img.shields.io/badge/Jul_24,_2022-black) ![](https://img.shields.io/badge/v0.4.1-REFACTOR-red)  
> `Alocated database files on diferent folders`

> ![](https://img.shields.io/badge/Jul_24,_2022-black) ![](https://img.shields.io/badge/v0.4.0-FEAT-blue)  
> `Added code coverage to pytest`

> ![](https://img.shields.io/badge/Jul_24,_2022-black) ![](https://img.shields.io/badge/v0.3.1-FIX-green)  
> `Invalid data tests passing with status_code 200`

> ![](https://img.shields.io/badge/Jul_23,_2022-black) ![](https://img.shields.io/badge/v0.3.0-FEAT-blue)  
> `Created some test cases`

> ![](https://img.shields.io/badge/Jul_21,_2022-black) ![](https://img.shields.io/badge/v0.2.0-FEAT-blue)  
> `Application able to run as a Docker container`

> ![](https://img.shields.io/badge/Jul_21,_2022-black) ![](https://img.shields.io/badge/v0.1.0-FEAT-blue)  
> `Both endpoints working`

tests/test_db.py
import pytest
import random

from .test_db_manager import db_connect


@pytest.fixture
def db():
    return db_connect()


def test_create_table(db):
    sql = '''
        CREATE TABLE IF NOT EXISTS test(
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            user_id INTEGER NOT NULL,
            name VARCHAR NOT NULL
        );
    '''

    cur = db.cursor()
    cur.execute(sql)
    db.commit()

    sql = '''
        SELECT name FROM sqlite_master
        WHERE type='table' AND name='test';
    '''
    cur.execute(sql)
    row = cur.fetchall()

    db.close()

    assert len(row) == 1


def test_insert_into_table(db):
    cur = db.cursor()

    for i in range(0, 5):
        sql = '''
            INSERT INTO test(user_id, name)
            VALUES(?, ?);
        '''

        cur.execute(sql, (random.randrange(100000, 999999), f"Tester_{i}"))

    db.commit()

    sql = '''
        SELECT * FROM test;
    '''

    cur.execute(sql)
    rows = cur.fetchall()

    db.close()

    assert len(rows) == 5


def test_drop_table(db):
    sql = '''
        DROP TABLE test;
    '''

    cur = db.cursor()
    cur.execute(sql)
    db.commit()

    sql = '''
            SELECT name FROM sqlite_master
            WHERE type='table' AND name='test';
        '''
    cur.execute(sql)
    row = cur.fetchall()

    db.close()

    assert len(row) == 0


tests/test_app.py
import os
import json

import pytest

from .test_db_manager import create_test_db, delete_tested_rows
from devgrid_challenge.src.app import app

create_test_db()


@pytest.fixture
def client():
    app.config['TESTING'] = True
    client = app.test_client()

    yield client


def test_env_exists():
    assert os.path.exists("../src/.env")


def test_post_with_valid_id(client):
    """
    Testing the main function with a valid id
    It must pass without raising any exception
    """
    user_id = "330104"

    url = "/weather"
    mimetype = "Application/json"

    headers = {
        "Content-Type": mimetype,
        "Accept": mimetype
    }

    body = {
        "user_id": user_id
    }

    result = client.post(url, data=json.dumps(body), headers=headers)

    response_body = json.loads(result.data)

    assert result.status_code == 200
    assert response_body == {"message": "All cities collected."}

    delete_tested_rows(user_id)


def test_post_with_existent_id(client):
    user_id = "123456"

    url = "/weather"
    mimetype = "Application/json"

    headers = {
        "Content-Type": mimetype,
        "Accept": mimetype
    }

    body = {
        "user_id": user_id
    }

    result = client.post(url, data=json.dumps(body), headers=headers)

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "Invalid 'user_id'. It must contain "
                                        "only numbers and be unique."}


def test_post_with_wrong_value_id(client):
    user_id = "1q2w3e"

    url = "/weather"
    mimetype = "Application/json"

    headers = {
        "Content-Type": mimetype,
        "Accept": mimetype
    }

    body = {
        "user_id": user_id
    }

    result = client.post(url, data=json.dumps(body), headers=headers)

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "Invalid 'user_id'. It must contain "
                                        "only numbers and be unique."}


def test_post_with_wrong_type_id(client):
    """
        Testing the main insertion function with an INVALID id
        It must throw an exception
        """
    user_id = {
        "id_type": "dict"
    }

    url = "/weather"
    mimetype = "Application/json"

    headers = {
        "Content-Type": mimetype,
        "Accept": mimetype
    }

    body = {
        "user_id": user_id
    }

    result = client.post(url, data=json.dumps(body), headers=headers)

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "Invalid 'user_id'. It must contain "
                                        "only numbers and be unique."}


def test_post_with_wrong_body(client):
    user_id = {
        "id_type": "dict"
    }

    url = "/weather"
    mimetype = "Application/json"

    headers = {
        "Content-Type": mimetype,
        "Accept": mimetype
    }

    body = {
        "test": user_id
    }

    result = client.post(url, data=json.dumps(body), headers=headers)

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "Missing param 'user_id'."}


def test_get_with_existent_id(client):
    user_id = 123456
    result = client.get(f"/weather?user-id={user_id}")
    assert result.status_code == 200


def test_get_with_unexistent_id(client):
    user_id = "000000"

    result = client.get(f"/weather?user-id={user_id}")

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "User identifier not found."}


def test_get_without_id(client):
    result = client.get("/weather")

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "Missing query param 'user-id'."}


def test_get_with_wrong_format_id(client):
    user_id = "1q2w3e"

    result = client.get(f"/weather?user-id={user_id}")

    response_body = json.loads(result.data)

    assert result.status_code == 400
    assert response_body == {"message": "Invalid 'user-id'. It must contain "
                                        "only numbers and be unique."}


tests/db_manager.py
import sqlite3
import random

from datetime import datetime


def db_connect():
    db = None
    try:
        db = sqlite3.connect("test_sqlite.db")
    except Exception as e:
        print(e)

    return db


def weather_exists():
    sql = '''
        SELECT name FROM sqlite_master
        WHERE type='table' AND name='weather';
    '''

    db = db_connect()
    cur = db.cursor()
    cur.execute(sql)
    row = cur.fetchall()

    if len(row) == 1:
        return True

    return False


def create_test_db():
    if not weather_exists():
        db = db_connect()
        cur = db.cursor()

        create_table = '''
                CREATE TABLE weather(
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    user_id INTEGER NOT NULL,
                    collected_at DATETIME NOT NULL,
                    city_info VARCHAR NOT NULL
            )
            '''

        cur.execute(create_table)

        for i in range(0, 3):
            user_id = 123456
            request_datetime = datetime.now()
            city_info = {
                "city_id": str(random.randrange(100000, 999999)),
                "temp": str(random.randrange(-70, 50)),
                "humidity": str(random.randrange(50, 100))
            }

            insert = f'''
                    INSERT INTO weather(user_id, collected_at, city_info)
                    VALUES({user_id}, datetime("{request_datetime}"), ?)
                '''

            cur.execute(insert, (str(city_info),))

        db.commit()


def delete_tested_rows(user_id):
    sql = f'''
        DELETE FROM weather
        WHERE user_id = {user_id}
    '''

    db = db_connect()
    cur = db.cursor()
    cur.execute(sql)
    db.commit()


src/app.py
import json
import os
import sqlite3
import requests

from flask import Flask, request
from dotenv import load_dotenv
from datetime import datetime

load_dotenv()
app = Flask(__name__)
app.config['JSON_SORT_KEYS'] = False


@app.route("/")
def index():
    return "Index"


@app.route("/weather", methods=['GET', 'POST'])
def weather():
    if request.method == "POST":
        request_data = request.json

        try:
            user_id = request_data["user_id"]
        except KeyError:
            return {"message": "Missing param 'user_id'."}, 400

        if verify_user_id(user_id):

            request_datetime = datetime.now()

            cities = get_appendix()
            try:
                for city_id in cities:
                    city_info = get_weather(city_id)
                    insert_weather(user_id, request_datetime, str(city_info))
            except Exception as e:
                return {"message": e}, 500

            return {"message": "All cities collected."}, 200
        else:
            return {"message": "Invalid 'user_id'. It must contain only "
                               "numbers and be unique."}, 400

    if request.method == "GET":
        user_id = request.args.get("user-id")

        if not user_id:
            return {"message": "Missing query param 'user-id'."}, 400

        try:
            int(user_id)
        except ValueError:
            return {"message": "Invalid 'user-id'. It must contain only "
                               "numbers and be unique."}, 400

        collected_cities = get_collected_cities(user_id)

        if not collected_cities:
            return {"message": "User identifier not found."}, 400

        return collected_cities, 200


def env_exisits():
    return os.path.exists("./.env")


def db_connect():
    db = None

    path = "../db"

    if not os.path.exists(path):
        os.makedirs(path)

    try:
        if app.config['TESTING']:
            db = sqlite3.connect("../db/test_sqlite.db")
        else:
            db = sqlite3.connect("../db/sqlite.db")
    except Exception as e:
        print(e)

    return db


def create_db():
    sql = '''
            CREATE TABLE IF NOT EXISTS weather(
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                user_id INTEGER NOT NULL,
                collected_at DATETIME NOT NULL,
                city_info VARCHAR NOT NULL
        )
        '''
    db = db_connect()
    cur = db.cursor()
    cur.execute(sql)
    db.commit()


def get_appendix():
    if app.config['TESTING']:
        appendix = os.getenv("APPENDIX_TEST").split(',')
    else:
        appendix = os.getenv("APPENDIX_A").split(',')

    return appendix


def verify_user_id(user_id):
    try:
        user_id = int(user_id)
    except ValueError:
        return False
    except TypeError:
        return False

    sql = "SELECT user_id FROM weather"
    db = db_connect()
    cur = db.cursor()
    cur.execute(sql)
    rows = cur.fetchall()

    for row in rows:
        if user_id == row[0]:
            return False

    return True


def get_weather(city_id):
    api_key = os.getenv('API_KEY')

    weather_api = "https://api.openweathermap.org/data/2.5/weather"
    query = f"?id={city_id}&appid={api_key}&units=metric"

    url = weather_api + query

    data = requests.get(url).json()

    return {
        "city_id": data['id'],
        "temp": data['main']['temp'],
        "humidity": data['main']['humidity']
    }


def insert_weather(user_id, request_datetime, city_info):
    sql = f'''
    INSERT INTO weather(user_id, collected_at, city_info)
    VALUES({user_id}, datetime("{request_datetime}"), ?)
    '''

    db = db_connect()
    cur = db.cursor()
    cur.execute(sql, (city_info,))
    db.commit()


def get_collected_cities(user_id):
    sql = f'''
        SELECT collected_at, city_info FROM weather
        WHERE user_id = {user_id}
        '''

    db = db_connect()
    cur = db.cursor()
    cur.execute(sql)
    rows = cur.fetchall()

    if len(rows) == 0:
        return False

    collected_info = {}

    cities_ids = get_appendix()
    progress = (len(rows) * 100) / len(cities_ids)

    collected_info["user_id"] = user_id
    collected_info["progress"] = f"{progress:.3g}%"

    cities = []
    for row in rows:
        city_info = row[1].replace("\'", "\"")
        city_info = json.loads(city_info)
        data_info = city_info
        data_info["collected_at"] = row[0]

        cities.append(data_info)

    collected_info["cities"] = cities
    return collected_info


if __name__ == "__main__":
    if not env_exisits():
        raise FileNotFoundError(
            ".env not found. Please make sure to get the file with the "
            "project maintainer and attatch to devgrid_challenge/src/ "
            "directory"
        )

    create_db()
    app.run(debug=True, host="0.0.0.0", port=5000)
